Merge in MAVLink Guide

hamishwillee opened the pull request:
This merges in the https://github.com/mavlink/mavlink/blob/master/doc/MAVLink2.md as proposed in #29 .
Wasn't sure about how up to date some parts are, so questions inline.

I am also thinking we should create a section "guide" for all the non-core topics - to control the size of the TOC.

hamishwillee commented on en/SUMMARY.md:
```diff
@@ -9,7 +9,7 @@
   * [Generate Source Files for ROS](getting_started/ros.md)
 * [MAVLink Versions](guide/mavlink_version.md)
-  * [MAVLink 2](guide/mavlink_2.md)
+* [MAVLink 2](guide/mavlink_2.md)
```
Note, now at same level. Propose to move this and versions topic into Guide, along with Serialisation etc. Will keep message definitions out of that guide.

hamishwillee commented on en/guide/mavlink_2.md:
```diff
@@ -1,28 +1,72 @@
+* [Extending existing MAVLink messages](#message_extensions)
+* [Variable length arrays](#variable_arrays) (*simplified implementation*).
 
+*MAVLink 2* bindings have been developed for C and Python (other languages are not implemented yet).
```
Supported languages section https://mavlink.io/en/#supported_languages says that only C and C++ 11 support Mavlink2. Is Python also supported?

hamishwillee replied:
@LorenzMeier Can you confirm the above?

LorenzMeier replied:
C and Python are definitely supported. All other languages have to be vetted one-by-one.

hamishwillee commented on en/guide/mavlink_2.md:
```diff
@@ -1,28 +1,72 @@
+
+For most users usage of the pre-generated C headers is recommended: https://github.com/mavlink/c_library_v2
+
+This will generate a set of C headers in the generator/C/include_v2.0 directory. 
```
This is copied from old doc, but I don't understand it - we're not generating anything here. Thoughts?

hamishwillee replied:
@LorenzMeier Thoughts on the above?

LorenzMeier replied:
Delete the one sentence, then it will be consistent.

hamishwillee commented on en/guide/mavlink_2.md:
```diff
@@ -1,28 +1,72 @@
+
+The major changes from an API perspective are:
+
+* you don't need to provide a message CRC table any more, or message length table. 
```
Is this relevant? Was this created for people, or something they had to do manually?

hamishwillee commented on en/guide/mavlink_version.md:
```diff
@@ -33,7 +33,30 @@ sequenceDiagram;
 Some popular legacy radios (e.g. the SiK radio series) operate in semi-transparent mode by injecting [RADIO_STATUS](../messages/common.md#RADIO_STATUS) messages into the MAVLink message stream. Per MAVLink spec these should actually emit a heartbeat with the same system ID and a different component ID than the autopilot to be discoverable. However, an additional heartbeat could be an issue for de ...
 
+
+## Negotiating Versions {#negotiating_versions}
```
SO I added this section from the guide. I think these go together - the first part explains how you can verify whether MAVLink 2 is supported or not, and the second is best practice guidance for systems to work out which version should be used for initial and subsequent comms. I think it is still useful?

hamishwillee commented on en/guide/message_signing.md:
```diff
@@ -0,0 +1,121 @@
+# Message Signing (Authentication)
```
I've split out the signing topics because I don't want to silo everything by version - except the stuff to do with "high level overview" and "version handshaking". Hope that makes sense.

LorenzMeier replied:
Makes sense, although v1 doesn't have signing and v3 will have asymmetric signing, so the information here is probably v2 specific.

LorenzMeier commented on en/guide/mavlink_2.md:
````diff
@@ -37,3 +81,38 @@ if (mavlink_parse_char(MAVLINK_COMM_0, buf[i], &msg, &status)) {
+
+```xml
+  <message id="152" name="MEMINFO">
+    <description>state of APM memory</description>
````
Please use a message from common.xml

hamishwillee replied:
@LorenzMeier Done.

hamishwillee commented on en/guide/mavlink_2.md:
```diff
@@ -37,3 +81,38 @@ if (mavlink_parse_char(MAVLINK_COMM_0, buf[i], &msg, &status)) {
 
+## Message Extensions {#message_extensions}
+
+The XML for messages can contain extensions that are optional in the protocol. 
```
@LorenzMeier I thought the purpose of extensions is to identify the new additions in MAVLink 2.0 so that they aren't generated into code for MAVLink 1. "The XML for messages can contain extensions that are optional in the protocol. " does not really capture that - can you explain?

The extensions thing seems a bit like a wasted opportunity to specify versioning.

hamishwillee replied:
Sure, discussion here: https://github.com/mavlink/mavlink/issues/919

With respect to the the decoding this is covered in the "The rules for extended messages" that follows the example. I will those up to make it more clear.

LorenzMeier commented on en/guide/message_signing.md:
````diff
@@ -0,0 +1,121 @@
+The solution adopted for MAVProxy is shown below:
+
+```
+if (msg.get_signed() and
````
Sorry, only saw this now - we need to replace this with an example from one of the maintained codebases. Or better: Generalize this to general MAVLink C++ code so we have all examples consistently in C++.

hamishwillee replied:
Can you point me to anything that uses this - quick search on `link_id`  shows no usage in the mavlink projects. 

Bigger deal is that I don't understand how this example can work. According to the text the problem is that "These multiple GCS instances will not be aware of each other, and so may choose the same link ID." However if your solution is to set the GCS ID from the autopilot link ID, and it is choosing its ID from the MAVLink channel number, then you're guaranteeing that all GCS connections to that autopilot WILL share the same id.

LorenzMeier commented on en/guide/message_signing.md:
```diff
@@ -0,0 +1,121 @@
+
+> **Note** Message signing is supported for [MAVLink 2](../guide/mavlink_2.md) (only).
+
+One of the key features of MAVLink2 is support for signing of messages. To enable signing in your application you will need to add some additional code. In particular you will need to add:
```
Missing space.

hamishwillee commented:
@LorenzMeier I have fixed the points that I can. There are some other questions in there for you:
1. https://github.com/mavlink/mavlink-devguide/pull/58/files#r188228530
2. https://github.com/mavlink/mavlink-devguide/pull/58/files#r188218856

hamishwillee commented:
@LorenzMeier There are some other questions there for you - but if you don't have time to answer them I'd rather get this in and merged than just sit here.

LorenzMeier commented:
I'm merging as is, but also provided answers to the questions you've raised, so we can address them in a follow-up PR.

hamishwillee commented:
Thanks! I fixed up the comments above in https://github.com/mavlink/mavlink-devguide/pull/60 (merged)

Deleted the old doc in https://github.com/mavlink/mavlink/pull/927
